File: axiom/discrete/sets/union_comprehension/emptyset.php
<?php
require_once '..\..\..\..\utility.php';
$i = 0;
$txt[$i++] = "\\(\\bigcup\\limits_{i=0}^{k} {x}_{i} = \\emptyset\\tag*{Eq[0]}\\)\\(\\forall_{\\substack{0 \\le i \\le k}}{{x}_{i} = \\emptyset}\\tag*{Eq[1]}\\)";
$txt[$i++] = "\\({x}_{j} = \\emptyset\\tag*{Eq[2]}\\)";
$txt[$i++] = "\\(\\exists_{{x}_{j}}{{x}_{j} \\neq \\emptyset}\\tag*{~Eq.paradox}\\)";
$txt[$i++] = "\\(\\exists_{{x}_{j}}{\\left|{{x}_{j}}\\right| > 0}\\tag*{~Eq.positive}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{k} {x}_{i}}\\right| = 0\\tag*{Eq.union_empty}\\)";
$txt[$i++] = "\\(\\bigcup\\limits_{i=0}^{k} {x}_{i} \\setminus {x}_{j} = \\emptyset\\tag*{Eq[3]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{k} {x}_{i} \\setminus {x}_{j}}\\right| = 0\\tag*{Eq[4]}\\)";
$txt[$i++] = "\\(\\left|{\\bigcup\\limits_{i=0}^{k} {x}_{i}}\\right| = \\left|{{x}_{j}}\\right| + \\left|{\\bigcup\\limits_{i=0}^{k} {x}_{i} \\setminus {x}_{j}}\\right|\\tag*{Eq[5]}\\)";
$txt[$i++] = "\\(0 = \\left|{{x}_{j}}\\right|\\tag*{Eq[6]}\\)";
$txt[$i++] = "\\(\\text{False}\\)";
render(__FILE__, $txt);
?>        

File: axiom/discrete/sets/union_comprehension/emptyset.py
from sympy.core.relational import Equality
from sympy.utility import plausible, Union
from sympy.core.symbol import Symbol, dtype
from sympy.sets.sets import Interval
from axiom import discrete
from sympy import S
from sympy.concrete.expr_with_limits import Forall

# given: Union[i](x[i]) = {}
# x[i] = {}


@plausible
def apply(given):
    assert given.is_Equality
    x_union, emptyset = given.args
    if emptyset != S.EmptySet:
        tmp = emptyset
        emptyset = x_union
        x_union = tmp
        assert emptyset == S.EmptySet

    assert x_union.is_UnionComprehension
    return Forall(Equality(x_union.function, S.EmptySet), *x_union.limits, given=given)


from sympy.utility import check


@check
def prove(Eq):
    i = Symbol('i', integer=True)
    k = Symbol('k', integer=True, positive=True)
    x = Symbol('x', shape=(k + 1,), dtype=dtype.integer)

    equality = Equality(Union[i:0:k](x[i]), S.EmptySet)

    Eq << apply(equality)

    j = Symbol('j', domain=Interval(0, k, integer=True))
    
    Eq << Eq[-1].limits_subs(i, j)
    
    Eq.paradox = ~Eq[-1]

    Eq.positive = Eq.paradox.apply(discrete.sets.inequality.strict_greater_than)

    Eq.union_empty = Eq[0].abs()

    Eq << Eq[0] - Eq.paradox.lhs

    Eq << Eq[-1].abs()

    Eq << Eq[-2].lhs.assertion()

    Eq << Eq[-1].subs(Eq[-2], Eq.union_empty)

    Eq << Eq[-1].subs(Eq.positive)


if __name__ == '__main__':
    prove(__file__)

